### PR TITLE
Feature/haz type in step impf

### PR DESCRIPTION
**Pull request by chahank**
Changes proposed in this PR:
- Make it obligatory for the sigmoid and step function methods to have haz_type defined
- Pass kwargs to the ImpactFunc init

### PR Author Checklist

- [x] Read the [Contribution Guide][contrib]
- [x] Correct target branch selected (if unsure, select `develop`)
- [x] Descriptive pull request title added
- [x] Source branch up-to-date with target branch
- [x] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [x] [Tests][testing] updated
- [x] [Tests][testing] passing
- [ ] No new [linter issues][linter]
- [x] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/main/CHANGELOG.md) updated

### PR Reviewer Checklist

- [x] Read the [Contribution Guide][contrib]
- [x] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/guide/Guide_Reviewer_Checklist.html) passed
- [x] [Tests][testing] passing
- [x] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Continuous_Integration_and_Testing.html
[linter]: https://climada-python.readthedocs.io/en/stable/guide/Guide_Continuous_Integration_and_Testing.html#3.C.--Static-Code-Analysis

### CODE REVIEW

**Review by leonie-villiger — COMMENTED** [review]
Thank you for implementing the keyword argument _haz_type_ into _ImpactFunc_´s classmethods _from_step_impf_ and _from_sigmoid_imp_ ! Please find some minor comments in the _Files changed_.

File: climada/entity/impact_funcs/base.py [hunk 1]
```diff
@@ -96,7 +96,7 @@ def __init__(
         self.mdd = mdd if mdd is not None else np.array([])
         self.paa = paa if paa is not None else np.array([])
 
-    def calc_mdr(self, inten):
+    def calc_mdr(self, inten: Union[float, np.ndarray]) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (chahank) [reply climada/entity/impact_funcs/base.py:99]: Exactly.

File: climada/entity/impact_funcs/base.py [hunk 2]
```diff
@@ -228,8 +248,12 @@ def from_sigmoid_impf(cls, intensity, L, k, x0, if_id=1):
...
             intensity value where f(x)==L/2
-        if_id : int, optional, default=1
+        haz_type: str
+            the reference string for the hazard (e.g., 'TC', 'RF', 'WS', ...)
+        impf_id : int, optional, default=1
             impact function id
+        kwargs :
+            keyword arguments passed to ImpactFunc()
 
         Return
```
[the remark this answers is not in the document]
Comment (chahank) [reply climada/entity/impact_funcs/base.py:258]: Yes, thanks!

File: climada/entity/impact_funcs/base.py
[the remark this answers is not in the document]
Comment (chahank) [reply climada/entity/impact_funcs/base.py:1]: That is a good question. I here propose to only change the specific class methods for step and for sigmoid functions. Changing the `__init__` would require some more thought about whether this is compatible with the rest of the CLIMADA code. Thus, to improve the code incrementally, I would first finish this pull request, and tackle the general question of haz_types for impact functions later. Maybe we should open an issue on that point to keep it on the agenda?

File: climada/entity/impact_funcs/test/test_base.py [hunk 3]
```diff
@@ -39,24 +39,29 @@ def test_calc_mdr_pass(self):
...
         """Check default impact function: step function"""
         inten = (0, 5, 10)
-        imp_fun = ImpactFunc.from_step_impf(inten)
+        imp_fun = ImpactFunc.from_step_impf(
+            intensity=inten, haz_type='TC', intensity_unit='m/s')
         self.assertTrue(np.array_equal(imp_fun.paa, np.ones(4)))
         self.assertTrue(np.array_equal(imp_fun.mdd, np.array([0, 0, 1, 1])))
         self.assertTrue(np.array_equal(imp_fun.intensity, np.array([0, 5, 5, 10])))
+        self.assertEqual(imp_fun.haz_type, 'TC')
+        self.assertEqual(imp_fun.intensity_unit, 'm/s')
```
[the remark this answers is not in the document]
Comment (chahank) [reply climada/entity/impact_funcs/test/test_base.py:48]: I decided to test only the explicit arguments of the method 
```
    def from_step_impf(
        cls, intensity: tuple[float, float, float], haz_type: str, mdd: tuple[float, float] = (0, 1),paa: tuple[float, float] = (1, 1), impf_id: int = 1, **kwargs):
```
and not the kwargs. I think this is ok like this, as one does not know how the kwargs might change in the future.

File: climada/entity/impact_funcs/test/test_base.py [hunk 4]
```diff
@@ -39,24 +39,29 @@ def test_calc_mdr_pass(self):
...
         inten = (0, 100, 5)
-        imp_fun = ImpactFunc.from_sigmoid_impf(inten, L=1.0, k=2., x0=50.)
+        imp_fun = ImpactFunc.from_sigmoid_impf(
+            inten, L=1.0, k=2., x0=50., haz_type='RF', name='sigmoid impf')
         self.assertTrue(np.array_equal(imp_fun.paa, np.ones(20)))
         self.assertEqual(imp_fun.mdd[10], 0.5)
         self.assertEqual(imp_fun.mdd[-1], 1.0)
         self.assertTrue(np.array_equal(imp_fun.intensity, np.arange(0, 100, 5)))
+        self.assertEqual(imp_fun.haz_type, 'RF')
+        self.assertEqual(imp_fun.name, 'sigmoid impf')
```
[the remark this answers is not in the document]
Comment (chahank) [reply climada/entity/impact_funcs/test/test_base.py:60]: Good point. I know added tests for the `id` but removed those for `unit` and `name` following the logic that only explicit arguments should be tested here.

File: climada/entity/impact_funcs/test/test_base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chahank) [reply climada/entity/impact_funcs/test/test_base.py:48]: I added a test for `id` too.

**Comment by chahank** [comment]
Thanks a lot for the thorough review! I think I addressed all points.

File: climada/entity/impact_funcs/base.py
[the remark this answers is not in the document]
Comment (leonie-villiger) [reply climada/entity/impact_funcs/base.py:1]: done; see: https://github.com/CLIMADA-project/climada_python/issues/686#issue-1646106299

**Pull request merged**